azure-storage-cpp-lite/src/blob/blob_client_wrapper.cpp

```diff
@@ -623,108 +630,101 @@ namespace microsoft_azure {
+                // Smoke check if the total size is known, otherwise - fail.
+                if (firstChunk.response().totalSize < 0) {
+                   errno = blob_no_content_range;
+                   return;
+                }
 
-                        {
-                            std::unique_lock<std::mutex> lk(ofs_mutex);
-                            writeCV.wait(lk, [&current, offset]() { 
-                                //std::cout << "offset: " << offset << std::endl;
-                                return current >= offset; });
-                            //if((unsigned long long)ofs.tellp() != offset)
-                            //{
-                            //    ofs.seekp(offset, std::ios_base::beg);
-                            //}
-                            std::ofstream ofs;
-                            if(offset == 0)
-                            {
-                                ofs.open(destPath, std::ofstream::out);
-                            }
-                            else
-                            {
-                                ofs.open(destPath, std::ofstream::out | std::ofstream::app);
-                            }
-                            ofs.write(os.str().c_str(), range);
-                            ofs.close();
-                            delete[] buffer;
+                // Get required metadata - etag to verify all future chunks and the total blob size.
+                const auto originalEtag = firstChunk.response().etag;
+                const auto length = static_cast<unsigned long long>(firstChunk.response().totalSize);
 
-                            if(offset + range > current)
-                            {
-                                current = offset + range;
-                            }
-                            //std::cout << "current: " << current << std::endl;
-                            writeCV.notify_all();
-                        }
+                // Create and truncate the target file.
+                auto fd = open(destPath.c_str(), O_CREAT|O_WRONLY, 0770);
+                if (-1 == fd) {
+                    return;
+                }
+                if (-1 == ftruncate(fd, length)) {
+                    close(fd);
+                    return;
+                }
+
+                // Persist the first chunk.
+                auto r = pwrite(fd, os.str().c_str(), firstChunk.response().size, 0);
+                while (r != static_cast<ssize_t>(firstChunk.response().size) && r != -1) {
+                    r = pwrite(fd, os.str().c_str(), firstChunk.response().size, 0);
+                }
+                if (-1 == r) {
+                    close(fd);
+                    return;
+                }
+                close(fd);
 
+                // Download the rest.
+                const auto left = length - firstChunk.response().size;
+                const auto chunk_size = std::max(DOWNLOAD_CHUNK_SIZE, (left + downloaders - 1)/ downloaders);
```

Comment:
This may make the chunk size too large. If the chunk size gets too large, we may also change the retry policy in get_chunk_to_stream_sync to resume from offset of left stream instead of start.

Reply:
why "too large"? this behaviour will not give bigger chunks than current single threaded download -) since we are not allocating a memory for the chunks, "too large" is not an issue, imho.

Reply:
I think the idea is to improve the case of a failure partway through downloading a very large file.  If we get a network issue when we've downloaded 90 MB of a 100 MB chunk, and the issue is temporary or retryable, better to only download the last 10 MB in the retry than the whole 100 MB.
Not a huge concern for me atm

Reply:
unfortunately retry logic is not implemented anyway - lite retries only on getting error status code and doesn't retry in case of network issues.